Live Chat - Implement moderation

akkadaska opened the pull request:
## Description

I implemented moderation of live chat, which is mentioned at issue #186.

- add `LiveChat.getItemMenu(item: ChatItemHasMenuEndpoint)`
  You can call this method with `ChatItemHasMenuEndpoint`: chat item which has `menu_endpoint` member such as 
  `LiveChatTextMessage`, and get all endpoints for moderating the chat item whose type is `Promise<Menu>`. You can call the 
moderation endpoint in the `Menu` instanse you got by following `LiveChat.selectButtonItem()` method.

- add `LiveChat.selectButtonItem(button: Button)`
  You can call chat moderation endpoint whose type is `Button` (or its sub class such as `MenuServiceItem`) with this method.
  The modelation endpoint comes from the response of  `LiveChat.getItemMenu()`, `LiveChatTextMessage.inline_action_buttons` or `LiveChatAutoModMessage.moderation_buttons`.

- fix some bugs

Here is a code example of new methods: Remove all LiveChatTextMessage that you have permittion for remove.
```typescript
liveChat.on('chat-update', async (chatAction: ChatAction) => {
  // only LiveChatTextMessage is filtered and assigned to liveChatTextMessage
  const menu = await liveChat.getItemMenu(liveChatTextMessage);
  const removeEndpoint = menu.items.filter((item) => {
      if (!item.is(MenuServiceItem)) {
        return false;
      }
      return item.as(MenuServiceItem).icon_type === 'DELETE';
    })[0];
  if (removeEndpoint) {
    liveChat.selectButtonItem(removeEndPoint.as(MenuServiceItem));
  }
}
```

Implement #186 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

akkadaska commented on src/parser/classes/NavigationEndpoint.ts:
```diff
@@ -230,12 +230,6 @@ class NavigationEndpoint extends YTNode {
       };
     }
-
-    if (data?.liveChatItemContextMenuEndpoint) {
```
This was duplicated.

akkadaska commented on src/parser/classes/NavigationEndpoint.ts:
```diff
@@ -307,7 +312,7 @@ class NavigationEndpoint extends YTNode {
     if (parse && result)
       return Parser.parseResponse(result.data);
 
-    return this.#call(actions, client);
```
In some cases this would execute #call twice.

akkadaska commented on src/parser/classes/livechat/items/LiveChatPaidSticker.ts:
```diff
@@ -42,7 +43,8 @@ class LiveChatPaidSticker extends YTNode {
     this.purchase_amount = new Text(data.purchaseAmountText).toString();
-    this.context_menu = new NavigationEndpoint(data.contextMenuEndpoint);
+    this.menu_endpoint = new NavigationEndpoint(data.contextMenuEndpoint);
+    this.context_menu = this.menu_endpoint;
```
Like others, this should be named menu_endpoint, not context_menu. To avoid breaking changes, I make both available.

LuanRT commented:
Awesome! I'll review it in a bit. Thank you for your work on this!

LuanRT commented on src/parser/youtube/LiveChat.ts:
```diff
@@ -180,6 +190,31 @@ class LiveChat extends EventEmitter {
     return data.actions.array().as(AddChatItemAction);
   }
 
+  async getItemMenu(item: ChatItemHasMenuEndpoint): Promise<Menu> {
```
Anyway we can have this method return a new object? Here's a small gist to demonstrate what I mean:

> ./youtube/ItemMenu.ts
```ts
//...

class ItemMenu {
  constructor(data: ParsedResponse) {
    this.#page = data;
  }
  
  /**
   * This should find and select the correct menu item.
   * So "title" would be either the label, text or icon_type.
   * If an item is not matched then throw an error. If it is matched
   * then call its endpoint.
   */
  async selectItem(title: string): Promise<ActionsResponse> {
    // ...
  }
  
  /**
   * Returns original InnerTube response.
   */
  async page(): ParsedResponse {
    return this.#page;
  }
}
```

Then `getItemMenu` in `LiveChat` at line 193 would return an `ItemMenu` object, which should take care of selecting items and calling their endpoints. Ex:

> ./youtube/LiveChat.ts
```js
//...
const response = await item.menu_endpoint.call(this.#actions, undefined, true);
return new ItemMenu(response);
```

This would simplify usage quite a lot, let me know what you think. Thanks!

akkadaska replied:
Thank you for the review. I have some worries.

1. When `ItemMenu.selectItem()` is called, to request to YouTube, we need to access LiveChat#actions. However, because `LiveChat#actions` is private member, we cannot acces it inside from `ItemMenu`. So some way to access is needed.
My idea is that LiveChat should have a method to call request to YouTube like call():
```typescript
class LiveChat {
  public async call(endpoint: NavigationEndpoint) {
    const response = await endpoint.call(this.#actions, undefined, true); 
    return response;
  }
}
```
and when `ItemMenu.selectItem()` is called, LiveChat.call() will be called inside from `ItemMenu.selectItem()`.

2. `LiveChatTextMessage.inline_action_buttons` and `LiveChatAutoModMessage.moderation_buttons` should be typed as ItemMenu, right?

3. Selecting item by its label, text or icon_type is simple especially deleting chat, but this may cause unstable works. `label` and `text` changes if another language is selected when `InnerTube.create()` is called. `icon_type` does not changes, but in some case, it is difficult to predict action from icon_type. For example, the icon_type for `hide this user` is `REMOVE_CIRCLE`.
So, I want additional way to call endpoint: receive endpoint instanses (currently typed `Button`) and select the endpoint by endpoint itself. I mean `LiveChat.selectButtonItem(button: Button)` is still needed.

LuanRT replied:
> When ItemMenu.selectItem() is called, to request to YouTube, we need to access LiveChat#actions. However, because LiveChat#actions is private member, we cannot acces it inside from ItemMenu. So some way to access is needed.
My idea is that LiveChat should have a method to call request to YouTube like call():

You don't need to create a new method ~
I think simply sharing the current `Actions` instance from `LiveChat` with the `ItemMenu` class would do, ex:
```js
class ItemMenu {
  #page;
  #actions;

  constructor(data: ParsedResponse, actions: Actions) {
    this.#page = data;
    this.#actions = actions;
  }
  // ...
}
```

Then:
```js
//...
const response = await item.menu_endpoint.call(this.#actions, undefined, true);
return new ItemMenu(response, this.#actions);
```
----

> LiveChatTextMessage.inline_action_buttons and LiveChatAutoModMessage.moderation_buttons should be typed as ItemMenu, right?

Nope, since `ItemMenu` will be a wrapper around actual menus it shouldn't interfere with individual parser types.

----

>  So, I want additional way to call endpoint: receive endpoint instanses (currently typed `Button`) and select the endpoint by endpoint itself. I mean `LiveChat.selectButtonItem(button: Button)` is still needed.

Yes, I agree. For that, you can simply expose items from the menu as public properties in the `ItemMenu` class and add an additional method or make `ItemMenu#selectItem` accept either a string or an item of type `Button`.

akkadaska replied:
OK! Thank you for your advice! Now I understand your idea.
I will work on to follow your design and request a review again.

akkadaska commented:
Now `LiveChat.getItemMenu(item)` returns a `Promise` of new class `ItemMenu`. 
We can select menu by `icon_type` of buttons like: 
```typescript
const itemMenu = await liveChat.getItemMenu(item);
itemMenu.selectItem('DELETE');
```
Also, we can get `Menu` instance by ItemMenu.menu(), which contains `Button` instances, and select item by `Button` instance like:
```typescript
const itemMenu = await liveChat.getItemMenu(item);
const targetButton = itemMenu.menu().items[0];
itemMenu.selectItem(targetButton);
```

I think we have one more task.
`LiveChatTextMessage.inline_action_buttons` or `LiveChatAutoModMessage.moderation_buttons` also contain `Button` instances to call. Currently, there are no appropriate way to call such `Button` items.
We can select item by `Button` using `ItemMenu.selectItem(button)`. However, to get `ItemMenu` instance, we need to call `liveChat.getItemMenu(item)`, which causes unnecessary request to YouTube.

My only idea is that `LiveChat` still should have a method to select item by `Button` instance. Please tell me your idea.

LuanRT commented:
> My only idea is that `LiveChat` still should have a method to select item by `Button` instance. Please tell me your idea.

I get what you're saying, the item menu is its own thing so that's why I suggested it to be a separate class. 

A generic button selector in the `LiveChat` class should be fine, unless we come up with a more elaborate way to deal with those additional buttons, but I can't really think of anything at the moment.

LuanRT commented on src/parser/youtube/ItemMenu.ts:
```diff
@@ -0,0 +1,59 @@
+    if (!endpoint.metadata.api_url) {
+      throw new InnertubeError('Response did not have an "api_url" property. The call may have failed.');
+    }
+    const response = await this.#actions.execute(endpoint.metadata.api_url, {...endpoint.payload, parse: true});
```
One question, is it possible to use the button endpoint's own `callTest` method or does it error out? I feel like the use of execute here might be redundant.

Ex:
```js
//...
const response = await endpoint.callTest(this.#actions, { parse: true });

akkadaska replied:
I have no special reason to use `execute`, so I change it to use `callTest`. I confirmed it works well.

LuanRT commented on src/parser/youtube/ItemMenu.ts:
```diff
@@ -0,0 +1,59 @@
+    if (!menu || !menu.is(Menu)) {
+      throw new InnertubeError('Response did not have an "live_chat_item_context_menu_supported_renderers" property. The call may have failed.');
+    }
+    this.#menu = menu.as(Menu);
```
How about we expose only the `items` prop of the menu in this class? I think it would make more sense since it's already named `ItemMenu`. That way, instead of having to rely on the `getter` (e.g; `itemMenu.menu().items`), they could be accessed through `itemMenu.items` which is a cleaner in my opinion.

akkadaska replied:
Yours is better. I fixed it.

akkadaska commented:
Finally I added `LineChat.selectButton()` to select moderation buttons in `LiveChatTextMessage.inline_action_buttons` or `LiveChatAutoModMessage.moderation_buttons`. And I fixed to follow your review.

I think all tasks about #202 is completed and it's ready to merge. Please review again.

LuanRT reviewed (approved):
Great! Lgtm too!
I'll check the rest of the code to make sure we're not missing anything and then merge it ~ 
And again, thank you!
